### ISSUE TITLE
[API Proposal] Add ReadOnlySequence<T>.TrySlice 

### ISSUE THREAD

**Issue by bsagal** [dotnet/corefx#38148]
When using ReadOnlySequence<T> if i want to see if i can get and get the `length` bytes stating from position `start` i need the following code:

```c#
    static bool TrySlice<T>(this ReadOnlySequence<T> seq, int start, int length, out ReadOnlySequence<T> result) {
        var slice = seq.Slice(start);
        if (slice.Length < length) {
            result = default;
            return false;
        }
        result = slice.Slice(length);
        return true;
    }
```

This code needs to allocate extra ReadOnlySequence, witch could be much better achieved in the sequence its self.

**Comment by stephentoub** [comment]
> needs to allocate extra ReadOnlySequence

What do you mean?  `ReadOnlySequence<T>` is a struct.

**Comment by bsagal** [comment]
What i meant to say is the slice method will need to be called multiple times, causing unnecessary calculations internally of the index and segment

**Comment by scalablecory** [comment]
Why are you calling `Slice` twice? Does this do what you want?

```
static bool TrySlice<T>(this ReadOnlySequence<T> seq, int start, int length, out ReadOnlySequence<T> result)
{
    if ((seq.Length - start) < length)
    {
        result = default;
        return false;
    }

    result = seq.Slice(start, length);
    return true;
}
```

**Comment by bsagal** [comment]
@scalablecory You are correct about my example, unforchantly most of the time use the slice overload where `start` is of type `SequencePosition` and there is currently no way to get the index from the `SequencePosition`

**Comment by scalablecory** [comment]
There are a few API proposals that might be useful here. A `long ReadOnlySequence<T>.GetIntegerPosition(SequencePosition)` doesn't sound unreasonable to me.

I'd encourage you to implement a quick benchmark (making use of `SequencePosition.GetObject()` and `SequencePosition.GetInteger()`) to see what difference this would make. This will help lend weight to your proposal. Note, using `ReadOnlySequence` directly for parsing can be very slow, so I would also take a look at your implementation to see if this is actually the best optimization you can make.
